python/robot_fingers/tasks/move_cube.py

```diff
@@ -0,0 +1,165 @@
+
+    # sample uniform position in circle (https://stackoverflow.com/a/50746409)
+    radius = _max_cube_com_distance_to_center * np.sqrt(random.random())
+    theta = random.uniform(0, 2 * np.pi)
+
+    # x,y-position of the cube
+    x = radius * np.cos(theta)
+    y = radius * np.sin(theta)
+
+    if difficulty == -1 or difficulty == 1:
+        # on the ground, random yaw
+        z = _CUBE_WIDTH / 2
+    elif difficulty == 2:
+        # in the air, random yaw
+        z = random.uniform(_min_height, _max_height)
+    else:
+        raise ValueError("Invalid difficulty %d" % difficulty)
+
+    position = np.array((x, y, z))
+
+    # random yaw angle (relative to current_orientation if given)
+    yaw = random.uniform(0, 2 * np.pi)
+
+    orientation = Rotation.from_euler("z", yaw).as_quat()
+    if current_orientation is not None:
+        orientation = orientation * current_orientation
+
+    return position, orientation
+
+
+def validate_goal(position, orientation):
+    """Validate that the given pose is a valid goal (e.g. no collision)
+
+    Args:
+        position:  Goal position.
+        orientation:  Goal orientation.
+
+    Returns:
+        (bool): True if the given goal pose is within the goal space and does
+            not collide with the static environment.
+    """
+    if len(position) != 3:
+        return False, "len(position) != 3"
+    if len(orientation) != 4:
+        return False, "len(orientation) != 4"
+    if np.linalg.norm(position[:2]) > _max_cube_com_distance_to_center:
+        return False, "Position is outside of the arena circle."
+    if position[2] <= _min_height:
+        return False, "Position is too low."
+    if position[2] >= _max_height:
+        return False, "Position is too high."
+
+    # even if the CoM is above _min_height, a corner could be intersecting with
+    # the bottom depending on the orientation
+    corners = get_cube_corner_positions(position, orientation)
+    min_z = min(z for x, y, z in corners)
+    if min_z < 0:
+        return False, "Position is too low."
+
+
+def evaluate_state(
+    goal_position, goal_orientation, actual_position, actual_orientation
+):
+    """Compute cost of a given cube pose.  Less is better.
+
+    Args:
+        goal_position:  Goal position (x, y, z).
+        goal_orientation:  Goal orientation as quaternion (x, y, z, q).
+        actual_position:  Goal position (x, y, z).
+        actual_orientation:  Goal orientation as quaternion (x, y, z, q).
+
+    Returns:
+        Cost of the actual pose w.r.t. to the goal pose.  Lower value means
+        that the actual pose is closer to the goal.  Zero if actual == goal.
+    """
+    # Use DISP distance (max. displacement of the corners)
+    goal_corners = get_cube_corner_positions(goal_position, goal_orientation)
+    actual_corners = get_cube_corner_positions(
+        actual_position, actual_orientation
+    )
+
+    disp = max(np.linalg.norm(goal_corners - actual_corners, axis=1))
```

Comment:
very elegant solution. i wonder, is the max the right thing here? or the mean? 

could this lead to counterintuitive behavior?

Reply:
This is the "displacement" metric as defined, for example, [here (p. 190, Example 5.6)](http://lavalle.pl/planning/bookbig.pdf).

Intuitively I would think that max is better than mean in cases of rotations (e.g. when the cube is rotated around one edge, such that the corners of this edge are still at their original position).  But this is just my gut feeling, I'm not aware of a systematic comparison.